mqtt/mqtt3/src/proto/deserialize.rs

```diff
@@ -0,0 +1,140 @@
+
+impl<'de, 'a> DeserializeSeed<'de> for SeededDeserializer<'a, Publication> {
+    type Value = Publication;
+
+    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        impl<'de, 'a> Visitor<'de> for SeededVisitor<'a, Publication> {
+            type Value = Publication;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+                write!(formatter, "an array of integers")
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+            where
+                A: SeqAccess<'de>,
+            {
+                let topic_name = seq
+                    .next_element()?
+                    .ok_or_else(|| de::Error::invalid_length(0, &self))?;
+                let qos = seq
+                    .next_element()?
+                    .ok_or_else(|| de::Error::invalid_length(1, &self))?;
+                let retain = seq
+                    .next_element()?
+                    .ok_or_else(|| de::Error::invalid_length(2, &self))?;
+                let payload = seq
+                    .next_element_seed(SeededDeserializer::<Bytes>(self.0, PhantomData))?
+                    .ok_or_else(|| de::Error::invalid_length(3, &self))?;
+
+                Ok(Publication {
+                    topic_name,
+                    qos,
+                    retain,
+                    payload,
+                })
+            }
+        }
+
+        deserializer.deserialize_seq(SeededVisitor::<Self::Value>(self.0, PhantomData))
+    }
+}
+
+impl<'de, 'a> DeserializeSeed<'de> for SeededDeserializer<'a, Bytes> {
+    type Value = Bytes;
+
+    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        impl<'de, 'a> Visitor<'de> for SeededVisitor<'a, Bytes> {
+            type Value = Bytes;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+                write!(formatter, "an array of integers")
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+            where
+                A: SeqAccess<'de>,
+            {
+                let payload: Vec<u8> = seq
+                    .next_element()?
+                    .ok_or_else(|| de::Error::invalid_length(0, &self))?;
+
+                let payload = self.0.payload_de_duper.de_dupe(payload);
+
+                Ok(payload)
+            }
+        }
+
+        deserializer.deserialize_seq(SeededVisitor::<Self::Value>(self.0, PhantomData))
+    }
+}
+
+struct PayloadDeDuper {
+    loaded_bytes: HashMap<u64, Bytes>,
+}
+
+impl PayloadDeDuper {
+    fn new() -> Self {
+        Self {
+            loaded_bytes: HashMap::new(),
+        }
+    }
+
+    fn de_dupe(&mut self, bytes: Vec<u8>) -> Bytes {
+        let hash = Self::calculate_hash(&bytes);
```

Comment:
why not just delegate it to hashmap itself?

Reply:
we propably can get away with HashSet
```rust
if let Some(bytes) = loaded_bytes.get(&bytes) {
  bytes.clone()
} else {
  loaded_bytes.insert(bytes);
  bytes.clone()
}
```

Reply:
Since the hash is only u64, collision is fairly likely due to birthday paradox

Reply:
yes, but exactly this HashSet implementation should do. If it is a collision - check for value equality